### PR TITLE
fix: ignore invalid and allow redundant role in aria-allowed-role

### PR DESCRIPTION
**Pull request by jkodu**
This PR tackles fix for ignoring invalid roles and allowing redundant roles in aria-allowed role. It also handles backwards compatibility for ARIA 1 vs 1.1 for certain roles like `combobox`, `spinbutton` and `searchbox`.

Closes issue:
- https://github.com/dequelabs/axe-core/issues/1110
- https://github.com/dequelabs/axe-core/issues/1116

## Reviewer checks

**Required fields, to be filled out by PR reviewer(s)**
- [x] Follows the commit message policy, appropriate for next version
- [x] Has documentation updated, a DU ticket, or requires no documentation change
- [x] Includes new tests, or was unnecessary
- [x] Code is reviewed for security by: @WilcoFiers

### CODE REVIEW

File: test/checks/aria/aria-allowed-role.js [hunk 1]
```diff
@@ -91,6 +91,36 @@ describe('aria-allowed-role', function() {
...
+		var node = document.createElement('input');
+		node.setAttribute('type', 'text');
+		node.setAttribute('role', 'searchbox');
+		fixture.appendChild(node);
+		assert.isTrue(
+			checks['aria-allowed-role'].evaluate.call(checkContext, node)
+		);
+	});
+
 	it('returns false when MENU has type context', function() {
```
Comment (isner) [inline test/checks/aria/aria-allowed-role.js]: Looks like the test description needs to be inverted to match the new expected result.

File: lib/commons/aria/get-role.js [hunk 2]
```diff
@@ -28,6 +29,11 @@ aria.getRole = function getRole(
 		}
 		return aria.isValidRole(role, { allowAbstract: abstracts });
 	});
+
+	if (segments) {
+		return validRoles;
+	}
```
Comment (WilcoFiers) [inline lib/commons/aria/get-role.js]: It's not at all obvious that `segments` wouldn't return the implicit role. I'm also not a fan of having this turn into a thing that can return either strings or arrays. You're trying to do two things with the same function. I suggest you keep the `getRoleSegments` function around, and have `getRole` call it to get the validRoles value instead.

File: lib/commons/aria/index.js [hunk 3]
```diff
@@ -1687,7 +1695,15 @@ lookupTable.role = {
 		nameFrom: ['author'],
 		context: null,
 		implicit: ['input[type="search"]'],
-		unsupported: false
+		unsupported: false,
+		allowedElements: [
```
Comment (WilcoFiers) [inline lib/commons/aria/index.js]: General question: How did we miss this, and how do we know we're not missing more stuff this time around?

File: lib/rules/aria-allowed-role-matches.js [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+
+// get role(s) as array(segments), with fallback and dpub
+const roleSegments = getRole(node, {
+	segments: true,
+	fallback: true,
+	dpub: true
+});
+
+// filter invalid roles
+const validRoles = roleSegments.filter(role => isValidRole(role));
```
Comment (WilcoFiers) [inline lib/rules/aria-allowed-role-matches.js]: getRole already filters out invalid roles, doesn't it? Couldn't you just do this:

```javascript
return getRole(node, { noImplicit: true, dpub: true }) !== null
```

File: test/integration/rules/aria-allowed-role/aria-allowed-role.html [hunk 5]
```diff
@@ -7,7 +7,6 @@
 <section id="pass-section-role-doc-bib" role="doc-bibliography"></section>
 <ul><li id='pass-li-role-doc-biblioentry' role="doc-biblioentry"></li></ul>
 <aside id='pass-aside-doc-example' role='doc-example'></aside>
-<div id='pass-div-has-any-role' role='divAnyRoleEvenInvalid'></div>
```
Comment (WilcoFiers) [inline test/integration/rules/aria-allowed-role/aria-allowed-role.html]: Instead of removing, I suggest you put in a valid role.

File: lib/commons/aria/index.js (hunk 3 above)
Comment (jkodu) [reply lib/commons/aria/index.js]: This is the holy grail source of truth - https://www.w3.org/TR/html-aria/ to construct these mappings.

There are 3 caveats with this:
1) Misinterpretation.
2) Conformance between ARIA1.0 and ARIA1.1, that was the case with https://github.com/dequelabs/axe-core/issues/1116. There were multiple ways to use a role. 1.1 mandates usage of role on parent, but backwards compatibility with 1.0 means role can be set on said element itself.
3) When https://www.w3.org/TR/html-aria/ is updated, we need a handle to confirm that things have not been changed, and adapt to new changes. Perhaps can watch/ github hook for the same and streamline the mapping.

To answer your specific Q, we missed this case due to an overlap of 1 and 2 above.

I am going through the table once more to ensure the mappings are right.

File: test/integration/rules/aria-allowed-role/aria-allowed-role.html (hunk 5 above)
Comment (jkodu) [reply test/integration/rules/aria-allowed-role/aria-allowed-role.html]: Done. Added it back again. This diff only watches this line, so comment is not going stale.

File: lib/commons/aria/get-role-segments.js [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+ * Returns all roles applicable to element in a list
+ *
+ * @method getRoleSegments
+ * @memberof axe.commons.aria
+ * @instance
+ * @param {Element} node
+ * @returns {Array} Roles list or empty list
+ */
+
+aria.getRoleSegments = function getRoleSegments(node) {
```
Comment (WilcoFiers) [inline lib/commons/aria/get-role-segments.js]: Either add tests for this, or don't move this into a commons. My preference would be the latter. I think you've made this PR much larger than it needed to be for the fix.

File: lib/commons/aria/get-element-unallowed-roles.js [hunk 7]
```diff
@@ -71,6 +37,12 @@ aria.getElementUnallowedRoles = function getElementUnallowedRoles(
 				return true;
 			}
 		}
+
+		// if role and implicit role are same, it is redundant - ignore
+		if (role === implicitRole) {
```
[the remark this answers is not in the document]
Comment (jkodu) [reply lib/commons/aria/get-element-unallowed-roles.js]: You are right, that flow path, only handles the edge case for `table>tr[role='row']`. I shall refactor.

File: lib/commons/aria/get-role-segments.js (hunk 6 above)
Comment (jkodu) [reply lib/commons/aria/get-role-segments.js]: There are tests (https://github.com/dequelabs/axe-core/pull/1118/files#diff-c01d2c9028bf4f6a16bace64d8ce59daR1) indeed . They were at the bottom of the PR.

I need to use `getRoleSegments` in 2 places, one within the `matches`, the other within the `check`, hence the abstraction to commons.

But believe I could use `getRole` in `matches` and I will move `getRoleSegments` out of commons.

**Comment by jkodu** [comment]
@WilcoFiers - reviewer checks were not filled up. Doing it on your behalf.
